ops/framework.py

```diff
@@ -984,16 +992,38 @@ def _reemit(self, single_event_path: Optional[str] = None):
                 if custom_handler:
                     event_is_from_juju = isinstance(event, charm.HookEvent)
                     event_is_action = isinstance(event, charm.ActionEvent)
-                    with self._event_context(event_handle.kind):
-                        if (
-                            event_is_from_juju or event_is_action
-                        ) and self._juju_debug_at.intersection({'all', 'hook'}):
-                            # Present the welcome message and run under PDB.
-                            self._show_debug_code_message()
-                            pdb.runcall(custom_handler, event)
-                        else:
-                            # Regular call to the registered method.
-                            custom_handler(event)
+                    synthetic = not event_is_from_juju and not event_is_action
+                    event_module = event.__class__.__module__
+                    if event_module.startswith('ops.'):
+                        # ops.charm.Events are re-exported through ops
+                        event_module = 'ops'
+                    event_class = f'{event_module}.{event.__class__.__qualname__}'
+                    # verbatim_name for ops and custom events, and kebab-hook-names for Juju events
+                    # NOTE: this is not the whole story, consider:
+                    # * some_relation-relation-broken
+                    # * workload_container_one-pebble-check-recovered
+                    event_name = (
+                        event_handle.kind if synthetic else event_handle.kind.replace('_', '-')
+                    )
+                    invocation = f'{event_name}: {observer_path}.{method_name}({event_class})'
+                    with tracer.start_as_current_span(invocation) as span:
+                        span.set_attribute('deferred', single_event_path is None)
+                        span.set_attribute('synthetic', synthetic)
+                        # FIXME: seems kinda redundant to have both event_class and event_name.
+                        span.set_attribute('event_class', event_class)
+                        span.set_attribute('event_name', event_name)
+                        span.set_attribute('handler', f'{observer_path}.{method_name}')
+                        # FIXME: for interesting events, may want to add interesting attributes:
```

Comment:
I think we would want to have all the key:values in the snapshot (if any) -- or the event arguments (other than the handle) if that was easier to get to, although in that case many are not strings so you'd need to str() them.

I think the only one that would be potentially sensitive might be the notice key, but it seems poor design if so (better to put something sensitive in a secret and put the secret ID in the notice data).